Cmake files added

JenniferBuehler opened the pull request:
This pull request only adds the cmake files. No changes to original source included yet.

JenniferBuehler commented:
It may be a bit confusing that there is a commit which talks about the locale settings still. This is because I originally had it in there. But have removed it in this branch which I used for the PR. So this PR only adds the CMake files and doesn't change anything in the source.

JenniferBuehler commented:
Ah, and another note: The cmake files still use catkin (though no other ROS packages). At some stage the catkin dependencies should be made optional, so that the CMakeLists can be used without catkin as well. Put this on the TODO list.

jvarley commented:
Overall, looks pretty good.  I will download and try it out this weekend.  Some of my thoughts:
1) I don't think this repo should have a package.xml, or any references to catkin/ros in the build including the commented out parts
2) The relevant code for the CGDB should be included
3) relevant code for PLY should be included
4) graspit/test/simple_test.cpp should be included and built as an additional executable.
4) has this been tested on windows and linux?

JenniferBuehler commented:
Sounds reasonable. I've updated the branch for my PR.

1. I've left the package.xml in there for now. This has the advantage that the package *can* (but does not *need* to) be compiled as catkin package. This is very convenient to build for people who are using it with ROS. For people who aren't, the package.xml is just a superfluous file, which should not disturb too much. I commented this in the description of the package.xml. If you really think we should remove it, that's ok. Just thought I'd keep it in there until everybody has voted on this.

2. CGDB is now included. I haven't tested related code because I don't know the code in this subject, so please give it a test run.

3. PLY now included as well. Same as for 2)

4. simple_test.cpp requires C++11. This has issues with compiling the other code. No major issues I think, but stuff that would need changing. I commented about this in the CMakeLists where add_executable for simple_test.cpp is commented out at the moment.

5. I don't use Windows. It was tested on Ubuntu 14.04. If you could test it on Windows, that would be awesome!

Note about the new CMakeLists.txt: It can be used with or without catkin. It detects whether catkin is used to compile, and if it is, it uses catkin commands. Otherwise, it doesn't. I have successfully compiled and run the simulator using cmake only as well.

JenniferBuehler commented:
*Side note:* If including the graspit library compiled with cmake only from the install directories (e.g. from an external package), a few changes in the original source are required: 
* change #include "src/Collision..." occurrences wit just "#include Collision..." and
* similarly, remvoing the "include/" from all #include "include/..."

The src/ and include/ directories should be added by the makefile. Such changes are only in the master branch of my fork so far. This commit still doesn't touch the sources.

jvarley commented:
So I am able to get this to build just fine when using cmake. 

When I use catkin_make, I get:

CMake Error at /opt/ros/hydro/share/catkin/cmake/catkin_package.cmake:291 (message):
  catkin_package() absolute include dir '/usr/include/qhull/qhull' does not
  exist
Call Stack (most recent call first):
  /opt/ros/hydro/share/catkin/cmake/catkin_package.cmake:98 (_catkin_package)
  graspit/CMakeLists.txt:692 (catkin_package)

should this line in CMakeMacros/FindQhull.cmake:
set(QHULL_INCLUDE_DIRS "${QHULL_INCLUDE_DIR}/qhull;${QHULL_INCLUDE_DIR}")
be:
set(QHULL_INCLUDE_DIRS "${QHULL_INCLUDE_DIR}")

If I change the above, then it builds fine with cmake and catkin_make.  I am on 12.04 not 14.04 on the machine I am currently testing this perhaps that accounts for the difference?

If I make this change, and then I build with catkin_make, It compiles just fine when it is alone. 

If I have a graspit plugin in my workspace, then catkin_make fails.  I have not been able to figure out exactly why yet.  As a minimal example, I make a workspace with
1) graspit
2) graspit_msgs https://github.com/CURG/graspit_msgs.git
3) graspit_moveit_plugin https://github.com/CURG/graspit_moveit_plugin.git

for the moveit plugin, I have to comment out everythin inside grasp_msg_publisher.cpp sendPickupRequest function this has to do with another PR.  
https://github.com/graspit-simulator/graspit/pull/9

With this workspace, caktin_make fails with:

> CMake Error at /home/jvarley/ros/temp_ws/build/graspit_msgs/cmake/graspit_msgs-genmsg.cmake:65 (add_custom_target):
>   add_custom_target cannot create target "graspit_msgs_generate_messages_cpp"
>   because another target with the same name already exists.  The existing
>   target is a custom target created in source directory
>   "/home/jvarley/ros/temp_ws/src/graspit_moveit_plugin".  See documentation
>   for policy CMP0002 for more details.
> Call Stack (most recent call first):
>   /opt/ros/hydro/share/genmsg/cmake/genmsg-extras.cmake:299 (include)
>   graspit_msgs/CMakeLists.txt:66 (generate_messages)
> 
> 
> CMake Error at /home/jvarley/ros/temp_ws/build/graspit_msgs/cmake/graspit_msgs-genmsg.cmake:124 (add_custom_target):
>   add_custom_target cannot create target
>   "graspit_msgs_generate_messages_lisp" because another target with the same
>   name already exists.  The existing target is a custom target created in
>   source directory "/home/jvarley/ros/temp_ws/src/graspit_moveit_plugin".
>   See documentation for policy CMP0002 for more details.
> Call Stack (most recent call first):
>   /opt/ros/hydro/share/genmsg/cmake/genmsg-extras.cmake:299 (include)
>   graspit_msgs/CMakeLists.txt:66 (generate_messages)
> 
> 
> CMake Error at /home/jvarley/ros/temp_ws/build/graspit_msgs/cmake/graspit_msgs-genmsg.cmake:183 (add_custom_target):
>   add_custom_target cannot create target "graspit_msgs_generate_messages_py"
>   because another target with the same name already exists.  The existing
>   target is a custom target created in source directory
>   "/home/jvarley/ros/temp_ws/src/graspit_moveit_plugin".  See documentation
>   for policy CMP0002 for more details.
> Call Stack (most recent call first):
>   /opt/ros/hydro/share/genmsg/cmake/genmsg-extras.cmake:299 (include)
>   graspit_msgs/CMakeLists.txt:66 (generate_messages)
> 
> 
> -- Configuring incomplete, errors occurred!
> 

But if I run it with the graspit-ros package rather than the graspit package, everything builds just fine.  The graspit_moveit_plugin does not generate any messages or have the targets that the error message is saying it does. I know the error doesn't mention anything to do with the graspit package, but it is causing the problem.   I am not sure what exactly is causing the problem.

JenniferBuehler commented:
Thanks for all the testing!

**1) Qhull error**

Yes the qhull error is most likely due to our different system configurations. The FindQHull.cmake is a copy of the PCL FindQhull.cmake (if I remember correctly)..

On my system it won't compile without the ``${QHULL_INCLUDE_DIR}/qhull``. This is because I have libqhull on my system... maybe you only have qhull? It ends up choosing libqhull directory instead of qhull, because that seems to be prioritized. If I *don't* add qhull to the include dir, it won't work, because qhull_a.h is not in libqhull.

To fix this, I removed the addition of qhull directory, so it'll compile on yours, and added a separate search for qhull_a.h so it'll compile on mine as well. I've updated the PR. Can you please try if this works fine on yours now?

**2) Message generation error**

I can't compile *graspit_moveit_plugin* with my fork of graspit:

```
Building CXX object graspit_moveit_plugin/CMakeFiles/ros_graspit_interface.dir/src/grasp_msg_publisher.cpp.o
/home/jenny/catkin_ws/src/graspit_moveit_plugin/src/planning_scene_msg_builder.cpp:3:42: fatal error: graspit_source/include/world.h: No such file or directory
 #include "graspit_source/include/world.h"
                                          ^
compilation terminated.
/home/jenny/catkin_ws/src/graspit_moveit_plugin/src/grasp_msg_publisher.cpp:4:47: fatal error: graspit_source/include/graspitGUI.h: No such file or directory
 #include "graspit_source/include/graspitGUI.h"
                                               ^
compilation terminated.

                                          ^
```

Because of the *graspit_source* reference I assume either the [CURG graspit-ros](https://github.com/CURG/graspit-ros) or the [graspit-simulator/graspit-ros](https://github.com/graspit-simulator/graspit-ros) are required to compile with cmake. So I've tried that, instead of using my fork. Unfortunately, neither of the graspit-ros packages work for me out-of-the box. 

With *CURG/graspit-ros*:
```
In file included from /home/jenny/catkin_ws/src/graspit_moveit_plugin/src/main.cpp:1:0:
/home/jenny/catkin_ws/src/graspit_moveit_plugin/include/ros_graspit_interface.h:43:28: fatal error: include/plugin.h: No such file or directory
 #include <include/plugin.h>
                            ^
compilation terminated.
```

There are more errors if I fix the individual issues, so I tried 

*graspit-simulator/graspit-ros*:

```
In file included from /home/jenny/catkin_ws/src/graspit_moveit_plugin/src/grasp_msg_publisher.cpp:8:0:
/home/jenny/catkin_ws/src/graspit-ros/graspit/./graspit_source/include/EGPlanner/searchState.h:34:22: fatal error: matvec3D.h: No such file or directory
 #include "matvec3D.h"
                      ^
compilation terminated.
In file included from /home/jenny/catkin_ws/src/graspit_moveit_plugin/src/planning_scene_msg_builder.cpp:3:0:
/home/jenny/catkin_ws/src/graspit-ros/graspit/./graspit_source/include/world.h:197:1: error: ‘signals’ does not name a type
 signals:
 ^
/home/jenny/catkin_ws/src/graspit-ros/graspit/./graspit_source/include/world.h: In member function ‘void World::tendonChange()’:
/home/jenny/catkin_ws/src/graspit-ros/graspit/./graspit_source/include/world.h:290:23: error: ‘emit’ was not declared in this scope
   void tendonChange(){emit tendonDetailsChanged();}
                       ^
/home/jenny/catkin_ws/src/graspit-ros/graspit/./graspit_source/include/world.h:290:28: error: expected ‘;’ before ‘tendonDetailsChanged’
   void tendonChange(){emit tendonDetailsChanged();}

< and so on>
```


I could fix the issues, but thought there's little point if I don't know which of the two packages you have been using, or if you have the graspit_source elsewhere in your system.

Please send me the details of your set-up so I can try to reproduce the error.

Anyway, two first thoughts: 

* If on my system any graspit-ros is required (given it didn't have errors) to
build graspit_moveit_plugin in the first place, but you can compile with my graspit fork *instead* of graspit-ros, then maybe you have other graspit sources on your system somewhere which graspit_moveit_plugin is using? Because it's not possible to compile *both* graspit-ros *and* my graspit fork (they are both called "graspit"). Is this intended?
* The error message suggests that it's a conflict between graspit_msgs message generation and graspit_moveit_plugin.  My cmake file does not do anything with graspit_msgs (I didn't have the package up until now). It also doesn't draw in any  dependencies which require either graspit_msgs or graspit_moveit_plugin...

As a first try I would edit line 155 in graspit_moveit_plugin/CMakeLists.txt and change it from

``add_dependencies(ros_graspit_interface graspit ${QT_DEFINITIONS})``

to

``add_dependencies(ros_graspit_interface graspit ${QT_DEFINITIONS} ${${PROJECT_NAME}_EXPORTED_TARGETS})``

I'm not sure this will make a difference at all but it's worth a shot. I've had message generation errors in the past which were often related to the exported targets (or lack of them).

JenniferBuehler commented:
P.S.: You could also try to comment out line 677 in my CMakeLists.txt:

``set (${PROJECT_NAME}_EXPORTED_TARGETS ${${PROJECT_NAME}_EXPORTED_TARGETS} ${catkin_EXPORTED_TARGETS})``

This line is not really needed to compile all my packages, I just added it for completeness so the cmake file can be used with or without catkin. Not sure if it does anything at all, because I'm not using add_dependencies() anywhere. But maybe the *_EXPORTED_TARGETS variables are used by the catkin build system internally...

jvarley commented on package.xml:
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0"?>
```
Can we fix the package xml to look more like:
https://github.com/graspit-simulator/graspit-ros/blob/master/graspit/package.xml

JenniferBuehler commented:
Ok I tried it again as now I'm testing on Indigo anyway. Here I could reproduce your error.
Turns out it was something very simple: 
graspit_moveit_plugin/package.xml was still missing the
<build_depend>graspit_msgs</build_depend>

It makes total sense because graspit_msgs was not processed *before* graspit_moveit_plugin in the cmake pipeline.

I also realized I had gotten something wrong in your last post, sorry about that: about you maybe having the graspit_source somewhere else. That's not needed of course if you have a working graspit-ros in your catkin_ws, and you compiled my graspit fork only with cmake or with no graspit_moveit_plugin in the catkin_ws... ouups. It must have been late ;)

Anyway, now it's starting to compile but I obviously get the same error like before from graspit_moveit_plugin because of the missing graspit_source. Only first line: 

``/home/jenny/catkin_ws/src/graspit_moveit_plugin/src/planning_scene_msg_builder.cpp:3:42: fatal error: graspit_source/include/world.h: No such file or directory
 #include "graspit_source/include/world.h"``

That's no surprise. I guess the #include directives would have to be changed in graspit_moveit_plugin, which would make sense anyway (i.e. take out all "graspit_source/" and "include/" from the includes and instead add required paths in the CMakeLists.txt)

jvarley commented on CMakeLists.txt:
```diff
@@ -0,0 +1,887 @@
+SET( QT_USE_QT3SUPPORT TRUE )
+find_package(Qt4 COMPONENTS QtCore REQUIRED)
+
+# include QT_USE_FILE is needed.
```
remove comment

jvarley commented on CMakeLists.txt:
```diff
@@ -0,0 +1,887 @@
+  ${GSRC}/include/Planner
+  ${GSRC}/src/Collision
+  ${GSRC}/cmdline
+  # following only has to be included because of Pr2Gripper!
```
remove comment

jvarley commented on CMakeLists.txt:
```diff
@@ -0,0 +1,887 @@
+set (GRASPIT_INCLUDEDIR_HEADERS
+    ${GSRC}/include/graspitParser.h 
+    ${GSRC}/include/barrett.h 
+	${GSRC}/include/body.h 
```
fix tabs

jvarley commented on CMakeLists.txt:
```diff
@@ -0,0 +1,887 @@
+	${GSRC}/src/profiling.cpp 
+	${GSRC}/src/pr2Gripper.cpp 
+	${GSRC}/src/m7.cpp 
+	${GSRC}/src/m7tool.cpp 
```
tabs

jvarley commented:
can we add to the commit to remove all the .pro files.

jvarley commented:
Hi Jennifer,
you said:
That's no surprise. I guess the #include directives would have to be changed in graspit_moveit_plugin, which would make sense anyway (i.e. take out all "graspit_source/" and "include/" from the includes and instead add required paths in the CMakeLists.txt)

What did you add to the CMakeLists.txt for this?

Thanks,
Jake

JenniferBuehler commented:
Hi Jake,

I edited CMakeLists.txt as you suggested, all makes sense. Can you check again if I got it right please?

To your question: I tried to compile graspit_moveit_plugin, but only got so far. Did the following changes
in graspit_moveit_plugin:

* grasp_msg_publisher.cpp:
    removed "graspit_source/include" and changed to:
    ```
    #include <graspit_msgs/Grasp.h>
    #include <graspitGUI.h>
    #include <world.h>
    #include <body.h>
    #include <robot.h>
    #include <EGPlanner/searchState.h>
    ```

* planning_scene_msg_builder.cpp:
    removed "graspit_source/include" and "include/" and changed to:
    ```
    #include <planning_scene_msg_builder.h>

    #include <world.h>
    #include <graspitGUI.h>
    #include <body.h>
    #include <triangle.h>
    ```


* CMakeLists.txt:
    added "include" directory as well:
    ```
    include_directories(
        include
        ${catkin_INCLUDE_DIRS}
        ${QT_INCLUDES})
    ```


But then I get stuck on indigo, I'll look into this, unless you already have some ideas?

```
Building CXX object graspit_moveit_plugin/CMakeFiles/ros_graspit_interface.dir/src/grasp_msg_publisher.cpp.o
In file included from /opt/ros/indigo/include/pluginlib/class_loader.h:301:0,
                 from /opt/ros/indigo/include/moveit/occupancy_map_monitor/occupancy_map_monitor.h:44,
                 from /opt/ros/indigo/include/moveit/planning_scene_monitor/planning_scene_monitor.h:46,
                 from /home/jenny/catkin_ws/src/graspit_moveit_plugin/include/planning_scene_msg_builder.h:7,
                 from /home/jenny/catkin_ws/src/graspit_moveit_plugin/src/planning_scene_msg_builder.cpp:1:
/opt/ros/indigo/include/pluginlib/class_loader_imp.h: In member function ‘std::string pluginlib::ClassLoader<T>::extractPackageNameFromPackageXML(const string&)’:
/opt/ros/indigo/include/pluginlib/class_loader_imp.h:210:41: error: no matching function for call to ‘TiXmlDocument::LoadFile(const string&)’
       document.LoadFile(package_xml_path);
                                         ^
/opt/ros/indigo/include/pluginlib/class_loader_imp.h:210:41: note: candidates are:
In file included from /opt/ros/indigo/include/srdfdom/model.h:45:0,
                 from /opt/ros/indigo/include/moveit/robot_model/robot_model.h:45,
                 from /opt/ros/indigo/include/moveit/robot_state/robot_state.h:41,
                 from /opt/ros/indigo/include/moveit/planning_scene_interface/planning_scene_interface.h:40,
                 from /home/jenny/catkin_ws/src/graspit_moveit_plugin/include/planning_scene_msg_builder.h:6,
                 from /home/jenny/catkin_ws/src/graspit_moveit_plugin/src/planning_scene_msg_builder.cpp:1:
/home/jenny/catkin_ws/src/graspit/tinyxml/tinyxml.h:1408:7: note: bool TiXmlDocument::LoadFile(TiXmlEncoding)
  bool LoadFile( TiXmlEncoding encoding = TIXML_DEFAULT_ENCODING );
       ^
/home/jenny/catkin_ws/src/graspit/tinyxml/tinyxml.h:1408:7: note:   no known conversion for argument 1 from ‘const string {aka const std::basic_string<char>}’ to ‘TiXmlEncoding’
/home/jenny/catkin_ws/src/graspit/tinyxml/tinyxml.h:1412:7: note: bool TiXmlDocument::LoadFile(const char*, TiXmlEncoding)
  bool LoadFile( const char * filename, TiXmlEncoding encoding = TIXML_DEFAULT_ENCODING );
       ^
/home/jenny/catkin_ws/src/graspit/tinyxml/tinyxml.h:1412:7: note:   no known conversion for argument 1 from ‘const string {aka const std::basic_string<char>}’ to ‘const char*’
/home/jenny/catkin_ws/src/graspit/tinyxml/tinyxml.h:1420:7: note: bool TiXmlDocument::LoadFile(FILE*, TiXmlEncoding)
  bool LoadFile( FILE*, TiXmlEncoding encoding = TIXML_DEFAULT_ENCODING );
       ^
/home/jenny/catkin_ws/src/graspit/tinyxml/tinyxml.h:1420:7: note:   no known conversion for argument 1 from ‘const string {aka const std::basic_string<char>}’ to ‘FILE* {aka _IO_FILE*}’
```

There seems to be an issue with the MoveIt! headers on indigo (included by planning_scene_msg_builder.h:7), not sure if I'm overlooking something here. I've just started to look into it now but have something else I need to do in the next hour or two...

jvarley commented:
Graspit has tiny xml included in it, which I would assume is an older version than the ones used in Moveit.

jvarley commented:
@JenniferBuehler 
I made several PRs to your branch to fix the tinyxml issue, allow tests to build correctly, and fix some indentations. Let me know if you see any problems with them.

JenniferBuehler commented:
Oh... just saw the message now after I pushed my fix as well... thanks for the hint, I didn't read the last line of the error message properly ;) Yes it was the graspit tinyxml header causing the issue. I removed it from the catkin_package include dirs and that fixed the issue. The only thing now could be that because the tinyxml graspit headers are not added as dependeceny, other graspit plugins etc could be missing them. Is any other package aside from graspit accessing tinyxml via the graspit dependencies?

I'll check out your changes, probably you ended up doing something similar?

Jenny

JenniferBuehler commented:
@jvarley @mateiciocarlie 

Hi, 
I see you have been making updates to your master repo which I would like to merge with my version in order to keep my work up-to-date with yours. I also would like to submit my next PR (about the abstract IVmgr), which actually doesn't incur many changes (and only small ones) to the original source. 

It would be easier for you if I based that PR on the most recent version of your master, and it would be easier for me if I can base that PR on a fresh pull from the master which already includes the CMake files.

So my question: Would you still like to merge this PR, and/or are there any more changes (aside from removing the new conflicts that popped up now) to happen before this PR can be merged? What's your plan/thoughts?

Thanks
Jenny

jvarley commented:
We should remove the package.xml, I missed it when purging catkin related code.  

Other than that, I don't see any reason not to pull this in.

@mateiciocarlie 
@JenniferBuehler

mateiciocarlie commented:
Looks great to me, thank you @jvarley and @JenniferBuehler for putting it all together! 

I do however want to take it for a test spin. I would also feel much better if it has been explicitly tested on Windows - did any of you get a chance to build on Windows?

JenniferBuehler commented:
Understandably, the more testing the better.

I haven't tried it on Windows. I don't have a win development environment set up at the moment - @jvarley , do you have one? I only have one little-maintained Windows install at the moment with windows 8.1 URGH probably not the best to test on ;-)

jvarley commented:
I have admittedly very little experience developing in Windows, nor do I have a windows dev environment either. I can give it a try tomorrow if no one else is willing....

JenniferBuehler commented:
I also have little Windows experience but will try it out too. Will first see if it works from win command line, if not it'll have to be cygwin as I don't have visual studio...

JenniferBuehler commented:
Alright, so it was never going to be a breeze... I opted for a MinGW/MSYS dev environment. 

I'm now stuck on not finding libBLAS, but that may be fixed quickly tomorrow. Anyway, meanwhile
I thought to share the steps I did to set it up, in case you find it useful @jvarley  because you were going to set it up today as well (or maybe it'll be easier with visual studio, there's actually a [free version](https://www.visualstudio.com/en-US/products/visual-studio-express-vs)).

Some instructions could also be useful to keep in an INSTALL.md, once tested properly.

**Set up MinGW and Qt**

1. Download MinGW toolchain (from MinGW-w64 project) already set up for Qt from [here](http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win32/Personal%20Builds/mingw-builds/4.8.2/threads-posix/dwarf/i686-4.8.2-release-posix-dwarf-rt_v3-rev3.7z/download) and extract it, e.g. to C:/MinGW. See also [this link on qt.io](https://wiki.qt.io/MinGW).

2. Download Qt 4 and run the installer (default destination is D:/Qt). I used the file [from qt.io](https://download.qt.io/archive/qt/4.8/4.8.6/) 
You have to locate the MinGW files previously extracted (see also [this useful resource](https://github.com/iat-cener/tonatiuh/wiki/Installing%20Qt%20For%20Windows)

3. Add MinGW *bin* directory to PATH

4. Download and install cmake for windows.
    *Notes for compiling with cmake:*
    In general, to compile with cmake from usual *build* directory, use
    ``cmake -G "MinGW Makefiles" ..``
    You may have to soft-link *ming32-make.exe* to *make.exe*.
    Weirdly, sometimes it complained about sh.exe, but re-running cmake helps.
    *Hint:* Pass -DCMAKE_PREFIX_PATH with a custom install directory for any dependencies installed in custom paths

**Install MSYS2**

* Download Msys2 [from here](https://msys2.github.io/) and execute installer.
* Open MSYS shell and add the MinGW *bin* directory to PATH (either export or .bashrc)
- *(optional)*: Install vim in from MSYS shell with ``pacman -S vim``


**Install basic dependencies**

1. lapack/blas: Install from source (see also [this link](http://icl.cs.utk.edu/lapack-for-windows/lapack/)), can be done [with this source](http://netlib.org/lapack/lapack.tgz) and cmake.
2. qhull : Compile from source [from github](https://github.com/qhull/qhull)

You could maybe also try pacman but not sure it gets the dev files.
``pacman -Ss lapack`` will list the specific name of the package (something with openblas) 
Then install with ``pacman -S <name>``
Same for qhull: find out package name with ``pacman -Ss qhull`` and install with ``-S``

**Install Coin and SoQt**

I mainly followed instruction given in [this link](https://github.com/iat-cener/tonatiuh/wiki/Installing-SoQt-For-Windows) (I also found [this](http://ascend4.org/Building_Coin3d_and_SoQt_on_MinGW) which may be useful)

- When installing coin, I needed extra flag in configure:  ``--build=x86_64-w64-mingw32``
(use name which ``gcc -dumpmachine`` outputs).
- For Coin, I had to to edit source code as suggested in the link's reference to the coin install, except the change in freetype.cpp wasn't necessary
- For Coin, also needed ``pacman -S diffutils``
- The Coin *bin* directory has to be in PATH
- Hint for SoQt: variable QTDIR has to be the directory where all the Qt includes are in

**Install bullet**

This was straight forward with cmake right [out of repository](https://github.com/bulletphysics/bullet3)


**Compile graspit**

Currently stuck at not finding libBLAS (though it's been installed), but gotta run now...

mateiciocarlie commented on CMakeLists.txt:
```diff
@@ -0,0 +1,962 @@
+set (GSRC ${PROJECT_SOURCE_DIR})
+
+set(PQP_COLLISION false CACHE PATH "if false, the in-build collision in graspit is used instead.")
+set(CGDB true CACHE PATH "compile CGDB sources as well")
```
Can we make this similar to the dynamics engine flag, where we have a COLLISION_ENGINE variable that can take multiple possible values, such as PQP_COLLISION or GRASPIT_COLLISION?

JenniferBuehler replied:
Sounds reasonable to me. When I continue with the Windows compilation test I can also add this change, if it's not been done before then (I'll probably get around to it by Wed /Thurs this week).

mateiciocarlie replied:
Great, thanks!

mateiciocarlie commented on CMakeLists.txt:
```diff
@@ -0,0 +1,962 @@
+    ${MOC_OUTFILES}
+)
+
+if (NOT SKIP_SHARED_LIB)
```
This part in general is quite confusing - not sure what to make of the BUILD_STATIC_LIB and especially the SKIP_SHARED_LIB flags. Is there a good reason to have multiple build options at this point?

JenniferBuehler replied:
I need the static lib because I am currently compiling the graspit source into my wrapper library - I wanted to avoid possible conflicts with other installed shared libraries of graspit (ie. libgraspit.so). It makes sense to do this while I still need the code from my fork.

SKIP_SHARED_LIB can be removed I guess. I  added it to speed up compilation time (takes forever when it compiles both static and shared afresh). I also use it when the shared lib is not needed, to avoid having it installed as well. A nicer approach would maybe be a single flag COMPILE_LIB_TYPE which can be set to "shared" or "static".

If you don't see any reason to include the option to compile a static library, it can be taken out of course. I could always maintain a separate CMakeLists.txt while I still need the static lib from my fork.

mateiciocarlie replied:
You are right, in the future it makes a lot of sense to build most of graspit as a lib, with the UI as an executable linking against it (or the command prompt version, etc.)

For now though, especially since this is a big change, I would prefer to keep things as simple as possible. If you don't mind maintaining your own slightly modified CMakeLists in your fork, I would prefer to go with a single build option...

mateiciocarlie commented:
Ran this on Linux as well and it worked great. I just added some small comments on the CMakeLists. 

Thanks for getting the Windows install procedure going - once we have that running end to end, this should be good to merge.

JenniferBuehler commented:
Alright, it was a long way to go with Windows (who would have thought). After a lot of trying around to find solutions for issues along the way, it finally works!!

![graspit](https://cloud.githubusercontent.com/assets/12842634/15098671/cb09de08-1542-11e6-9f1a-3cd42c4d3876.png)

I had to adjust a few details in the CMakeModules and the CMakeLists.txt, incl. removal of SKIP_SHARED and COMPILE_STATIC, as per comments above. I left a few comments in there for future reference, I hope this is ok? Could save some hassle in future.
The most tricky bit was probably having to change the way plugin.h/.cpp were handling the loading of libraries (see explanation (*) below). I don't know which is the best plugin to test it again - @jvarley could you maybe test any plugin to see if it still works (ie. it still loads the libraries), since you're working on plugins anyway?

I have added a [Install.MinGW.md](https://github.com/JenniferBuehler/graspit/blob/master/INSTALL.MinGW.md) to my master branch if you would like to try it out. Let me know if you'd like me to add this file to this PR as well.

@jvarley @mateiciocarlie  I have still not removed the PQP_COLLISION flag as I'm not sure how this works: Is this part of DYNAMICS_ENGINE=="GRASPIT_DYNAMICS" ? Ie. could we use DYNAMICS_ENGINE *instead* of PQP_COLLISION, or should it exist as separate flag?

**(*) Side note:** Why did the dynamic linking have to be changed in plugin.h/.cpp?
On MinGW/MSYS, I had problems with undefined references to DL functions such, as *dlopen()* etc. This only happened with the compiler (g++ 4.8.2) coming with the Qt MinGW build (downloaded in first step as per INSTALL.MinGW.md). That's the one compiler that should be used after all (aside from that, I had problems with others I tried as well, especially in combination with cmake). This issue could be handled by changing plugin.h/.cpp to use the windows libary, ie. LoadLibrary() etc. Worked fine on MinGW. The non-windows version should still work as before, but should be tested again.

JenniferBuehler commented:
One thing I forgot to say about the (*) side note posted above.
In case you wonder whether it may have been related to missing libraries: I tested that. I broke it down to a simple test.cpp (with one test call to dlclose) linking explicitly to the existing libdl.so.a. It still had undefined references, when compiled with the said "Qt-compatible" g++.exe. However when compiled with other mingw/msys g++.exe's, it wasn't even required to specify -ldl --- it compiled fine without.
I tried a lot of things, but then finally worked around this issue by using the windows dynamic library loaders. I think it's definitely good to keep it compatible with the mingw build which is "officially" tested with Qt.

jvarley commented:
I think a good plugin to look at is the openclose plugin inside graspit as it includes a good number of GraspIt! headers.  It currently does not work with the cmake graspit version.  I made a PR to this branch with plugin fixes for the demo plugins that come with graspit.  The only change that I don't really like is how I fixed the .pro file to allow the plugin to find the ui_* headers, but I am not sure of a better way to do this.  I think it will be helpful to write a findGraspit.cmake sometime soon.

If you want to run the plugin
cd graspit
export GRASPIT=$PWD
cd graspit/plugins/openclose
qmake openClosePlugin.pro
make -j5
cd graspit/plugins/openclose/lib
export GRASPIT_PLUGIN_DIR=$PWD
cd graspit
./build/graspit_simulator -p libopenClosePlugin

JenniferBuehler commented:
Thanks for checking this. I will test it as soon as I can.
Was it working with the cmake version before the changes I made for Windows?

jvarley commented:
@JenniferBuehler 

I have never done anything with the PQP_COLLISION flag. My understanding of it is that Graspit can use either PQP for collision detection, or graspit's own collision detection.  Except for when Bullet is in use, and dynamics are running, then it uses bullet's collision detection.

jvarley commented:
I never tried the plugins prior to the windows changes.

JenniferBuehler commented:
Ok, thanks. I hope to get the time to look at it one of the next days.
Out of curiosity: Is it seg-faulting, or printing an error?

So to confirm with @mateiciocarlie, we could use the following DYNAMICS_ENGINE flags and put it all under one hood:

* GRASPIT_DYNAMICS  (using graspit dynamics *and* collision detection)
* GRASPIT_DYNAMICS_PQP (using graspit dynamics and PQP collision detection)
* BULLET_DYNAMICS (using bullet dynamics and bullet collision detection)

Or am I missing anything? Is it possible to use *no* dynamics, but still use bullet, but with PQP collision detection *instead* of bullet collision detection?

jvarley commented:
I think you got it.  I cannot currently use bullet without dynamics. What is it printing?

JenniferBuehler commented:
I haven't tried it yet, just working on something else. What I was asking is what you mean with plugins  not currently working with the cmake version. What is going wrong: Segfault? Error?

mateiciocarlie commented:
The choice of dynamics and the choice of collision engine should be independent, even though some dynamics engines will actually provide both. Also, one can choose to use a different collision engine and just never use the dynamics engine at all - a lot of GraspIt's usefulness actually comes from the static simulations.

So perhaps we can have two independent flags, e.g. DYNAMICS_ENGINE and COLLISION_ENGINE?

JenniferBuehler commented:
Sounds reasonable, so confirming:

DYNAMICS_ENGINE:
* GRASPIT_DYNAMICS
* BULLET_DYNAMICS

COLLISION_ENGINE
* GRASPIT_COLLISION
* BULLET_COLLISION

Choosing GRASPIT_DYNAMICS and BULLET_COLLISION, and probably other combinations, would be invalid though? So checking for inconsistencies may get a little convoluted, but that's not avoidable if making 2 variables of it...

jvarley commented:
It didn't compile unless I made the changes from the PR I sent you.  With the PR, it runs fine.

mateiciocarlie commented:
Agreed! COLLISION_ENGINE should also get the PQP_COLLISION option though.

JenniferBuehler commented:
Oh, of course, forgot PQP_COLLISION. So no error checking in cmakelists.txt?

@jvarley: Ok so the plugin does work with the new windows changes, after your PR?

jvarley commented:
yes

JenniferBuehler commented:
YAY!!!

jvarley commented:
lol, I know right.  Thanks for checking out windows, that would have taken me forever.

JenniferBuehler commented:
It did take me forever too. Compiling on Windows is a nightmare. Always you think you're almost there, and then... you aren't. I hope to not see Windows for a long, long while after this is done ;)

JenniferBuehler commented:
@jvarley You can use [this Findgraspit.cmake](https://github.com/JenniferBuehler/graspit-pkgs/blob/master/grasp_planning_graspit/CMakeMacros/Findgraspit.cmake) to start with, it can probably be improved.

jvarley commented:
I will checkout the Findgraspit.cmake.  

I just made a pr for graspit-ros:
https://github.com/graspit-simulator/graspit-ros/pull/6

This switches it over to work with graspit built from cmake.  It still has graspit as a submodule, I don't know if we want to keep it that way, or switch over to using a Findgraspit.cmake similar to gazebo, pcl etc.  

I am still having a small problem, where it places graspit in build rather than devel, I am trying to fix it at the moment, let me know if you guys know how. 

Once this is fixed, I think we are ready to pull in everything.

jvarley commented:
@mateiciocarlie @JenniferBuehler 

I think we are ready to pull in both this PR, and 
https://github.com/graspit-simulator/graspit-ros/pull/6
Then we can up the submodule for graspit-ros to point at this version of GraspIt!.

Best of my knowledge, we currently work 
- with/without bullet
- windows/linux
- with non ros plugins
- with ros plugins and new version of graspit-ros

Any concerns?

-Jake

JenniferBuehler commented:
@jvarley if you build graspit with cmake (not with catkin), it will always put it in the build folder. It is catkin which changes the install destination to the devel directory. You could maybe try setting CMAKE_INSTALL_PREFIX manually? Are you trying to compile graspit from within the catkin workspace (eg. pulling it in as external cmake source)?

jvarley commented:
I fixed it, I just need to move the catkin_package linie above the add_subdirectory line in the graspit-ros cmakelist.txt

jvarley commented:
Currently graspit is a submodule to graspit-ros.  I think at some point I would prefer to make it an external cmake source, but I need to talk to Matei about that, and I think it can be left for another day. At the moment, I think we are good to go.

mateiciocarlie commented:
Hooray! Merged.

Thanks for all the contributions here - it was s big one!
